apptools/lru_cache/lru_cache.py

```diff
@@ -77,54 +71,17 @@ def __len__(self):
             return len(self._cache)
 
     def __getitem__(self, key):
-        PREV, NEXT = 0, 1   # names for the link fields
-        # Size limited caching that tracks accesses by recency
         with self._lock:
-            link = self._cache[key]
-            if link is not None:
-                # Move the link to the front of the circular queue
-                link_prev, link_next, _key, result = link
-                link_prev[NEXT] = link_next
-                link_next[PREV] = link_prev
-                last = self._root[PREV]
-                last[NEXT] = self._root[PREV] = link
-                link[PREV] = last
-                link[NEXT] = self._root
-                return result
+            return self._renew(key)
 
     def __setitem__(self, key, result):
-        PREV, NEXT, KEY, RESULT = 0, 1, 2, 3   # names for the link fields
         try:
-            with self._lock:
-                if len(self) == self.size:
-                    # Use the old root to store the new key and result.
-                    oldroot = self._root
-                    oldroot[KEY] = key
-                    oldroot[RESULT] = result
-                    # Empty the oldest link and make it the new root.
-                    # Keep a reference to the old key and old result to
-                    # prevent their ref counts from going to zero during the
-                    # update. That will prevent potentially arbitrary object
-                    # clean-up code (i.e. __del__) from running while we're
-                    # still adjusting the links.
-                    self._root = oldroot[NEXT]
-                    oldkey = self._root[KEY]
-                    oldresult = self._root[RESULT]
-                    self._root[KEY] = self._root[RESULT] = None
-                    # Now update the cache dictionary.
-                    del self._cache[oldkey]
-                    if self.cache_drop_callback is not None and oldkey != key:
-                        self.cache_drop_callback(oldkey, oldresult)
-                    # Save the potentially reentrant cache[key] assignment
-                    # for last, after the root and links have been put in
-                    # a consistent state.
-                    self._cache[key] = oldroot
-                else:
-                    # Put result in a new link at the front of the queue.
-                    last = self._root[PREV]
-                    link = [last, self._root, key, result]
-                    last[NEXT] = self._root[PREV] = self._cache[key] = link
-            return None
+            self._cache[key] = result
```

Comment:
Everything up to calling the callback should be under the lock. `OrderedDict` is not thread safe (http://bugs.python.org/issue11875#msg134089), so every addition and removal should be locked.

Beyond that, this contains a possible race condition:
1) Thread 1 just cached a value and execution just finished line 81.
2) Thread 2 just finished line 79. The key it is using is the current LRU.
3) Thread 1 executes line 82, popping the key and value just set by Thread 2.
4) Thread 2 eventually gets to line 80 and *crash*.